Sources/BSON/BSONCode.swift

```diff
@@ -88,6 +88,55 @@ extension BSONCode: BSONValue {
 }
 
 extension BSONCodeWithScope: BSONValue {
+    /*
+     * Initializes a `BSONCode` from ExtendedJSON.
+     *
+     * Parameters:
+     *   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for `Code`.
+     *   - `keyPath`: an array of `String`s containing the enclosing JSON keys of the current json being passed in.
+     *              This is used for error messages.
+     *
+     * Returns:
+     *   - `nil` if the provided value is not a `String`.
+     *
+     * Throws:
+     *   - `DecodingError` if `json` is a partial match or is malformed.
+     */
+    internal init?(fromExtJSON json: JSON, keyPath: [String]) throws {
+        switch json {
+        case let .object(obj):
+            // canonical and relaxed extended JSON
+            guard
+                let code = obj["$code"],
+                let scope = obj["$scope"]
+            else {
+                return nil
+            }
+            guard obj.count == 2 else {
+                throw DecodingError._extendedJSONError(
+                    keyPath: keyPath,
+                    debugDescription: "Expected only \"$code\" and \"$scope\" keys, got: \(obj.keys)"
+                )
+            }
+            guard let codeStr = code.stringValue else {
+                throw DecodingError._extendedJSONError(
+                    keyPath: keyPath,
+                    debugDescription: "Could not parse `BSONCodeWithScope` from \"code\": \"\(code)\"," +
+                        " input must be a string."
+                )
+            }
+            guard let scopeDoc = try BSONDocument(fromExtJSON: scope, keyPath: keyPath + [codeStr]) else {
```

Comment:
`keyPath` is reserved for keys only, so we don't want to be appending the code value here.

In general, the keyPath just tracks where in the output BSON document we are when we encounter an error in the input JSON. This allows us to tell the user where to look for an error occurring.

For example, given input JSON:
```json
{ 
    "foo": { "$numberInt": "12" },
    "bar": { "baz": { "$code": "console.log(x)", "$scope": { "x": 5 } } }
}
```
the output BSON document would look like:
```javascript
{
    "foo": 12,
    "bar": { "baz": CodeWithScope(...) }
}
```
So the only values that should ever end up in keyPath are `["foo"]` or `["bar"]` or `["bar, "baz"]`, since those are the only _actual keys_. All the other keys encountered ("$code", "$scope", etc.) are part of the _values_ themselves.

Now, if the value for the `CodeWithScope` in this example was malformatted and we threw an error (say the scope was incorrect), `keyPath` would be `["bar", "baz"]` and our error would then include the string "bar.baz", so they know what the issue was. 

So, getting back to the original comment, `codeStr` here refers to `"console.log(x)"` in this example, so we definitely don't want to be putting that in the keyPath. More specifically, we only ever need to modify the keyPath when descending recursively from within a `BSONDocument` or `[BSON]` initializer.

Reply:
Correction: I realize now that when we recursively enter the scope `BSONDocument`, we actually do need to modify the keyPath in the event that it has a decoding error somewhere down the line within the scope document. This is probably what you were going for originally.

For example, for the following badly formatted JSON:
```javascript
{ "mycode": { "$code": "console.log(5)", "$scope": { "x": 2, "y": { "$numberInt": "ERROR" } } } }
````
we don't want this to give us `DecodingError("mycode.y: ...")`, we want it to say `DecodingError("mycode.$scope.y: ...")`.

So, most of what I said still holds, but you are correct that we need to append to the keyPath here--just instead of appending `codeStr` we append `"$scope"`. To generalize this rule: we append to the keypath whenever we recurse via `BSON`, `Array`, or `BSONDocument`. 

Sorry for the confusion!